### PR TITLE
chore: Instrument modal for funnels

### PR DESCRIPTION
**Pull request by michaeldowseza**
### Description

This change adds funnel support for the Modal component. This allows the potential collection of user interactions through deletion flows within Modals.

Quip: tiy2A6a3X2qu

#### Change summary
- Adds funnel instrumentation to Modals
- Add analyticsMetadata with `delete` type to Modal

### How has this been tested?

- Unit tests added for Modals

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/internal/analytics/__tests__/mocks.ts [hunk 1]
```diff
@@ -48,3 +48,25 @@ export function mockInnerText() {
     afterEach(() => delete (HTMLElement.prototype as Partial<HTMLElement>).innerText);
   }
 }
+
+export function mockGetBoundingClientRect() {
```
Comment (michaeldowseza) [inline src/internal/analytics/__tests__/mocks.ts:52]: Was re-used in a couple of tests. Refactored into this helper mock function.

File: src/internal/analytics/components/analytics-funnel.tsx [hunk 2]
```diff
@@ -384,6 +394,48 @@ const InnerAnalyticsFunnelStep = ({
...
+    }
+
+    if (mounted) {
+      return;
+    }
+
+    const stepName = getTextFromSelector(stepNameSelector);
+    const handler = setTimeout(() => {
+      if (funnelState.current !== 'cancelled') {
+        FunnelMetrics.funnelStepComplete({
```
[the remark this answers is not in the document]
Comment (michaeldowseza) [reply src/internal/analytics/components/analytics-funnel.tsx:410]: As you mention this code is reached by setting `visible={false}`. Developers don't need to unmount the component if they don't want and so the clean up effect is not run in this case. The effect below this one handles the mount/unmount of the component.

So the cases we need to handle are
```
{visible && <Modal visible={true} />} // Handled by mount/unmounted
{visible && <Modal visible={visible} />} // Handled by mount/unmounted, component unmount before effect of visible property change is executed
{<Modal visible={visible} />} // Handled by this new effect 
```

[the remark this answers is not in the document]
Comment (connorlanigan) [reply src/internal/analytics/components/analytics-funnel.tsx:410]: Ah, I see. Thank you for the explanation!

File: src/modal/__tests__/analytics.test.tsx [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    act(() => void jest.runAllTimers());
+    expect(FunnelMetrics.funnelComplete).toHaveBeenCalledTimes(1);
+    expect(FunnelMetrics.funnelComplete).toHaveBeenCalledWith(
+      expect.objectContaining({
+        funnelInteractionId: expect.any(String),
+      })
+    );
+  });
+
+  test('sends funnelSuccessful when primary button in footer is clicked and modal is dismissed', () => {
```
Comment (connorlanigan) [inline src/modal/__tests__/analytics.test.tsx:136]: We could potentially also test that the same event is emitted when the primary button is clicked and the modal is _unmounted_

**Review by connorlanigan — DISMISSED** [review]

File: src/modal/index.tsx [hunk 4]
```diff
@@ -2,17 +2,88 @@
...
 import { applyDisplayName } from '../internal/utils/apply-display-name';
 import { ModalProps } from './interfaces';
-import InternalModal from './internal';
+import InternalModal, { InternalModalAsFunnel } from './internal';
+
+import styles from './styles.css.js';
 
 export { ModalProps };
 
+function ModalWithAnalyticsFunnel({
```
[the remark this answers is not in the document]
Comment (michaeldowseza) [reply src/modal/index.tsx:21]: How about moving it into internal? We have a bit of mix at the moment between some of the other components so no real set standard.

**Review by taheramr — DISMISSED** [review]

File: src/modal/index.tsx [hunk 5]
```diff
@@ -2,17 +2,88 @@
...
+      props: { size, disableContentPaddings: props.disableContentPaddings },
+      metadata: {
+        hasResourceType: Boolean(analyticsMetadata?.resourceType),
+        hasInstanceIdentifier: Boolean(analyticsMetadata?.instanceIdentifier),
+      },
+    },
+    analyticsMetadata
+  );
+
+  if (!isInFunnel) {
```
[the remark this answers is not in the document]
Comment (michaeldowseza) [reply src/modal/index.tsx:76]: As discussed this is only going to apply to non nested funnels.

**Review by amrtaher1234 — APPROVED** [review]

**Review by connorlanigan — APPROVED** [review]

**Review by taheramr — APPROVED** [review]

**Pull request merged**